src/validation.cpp

```diff
@@ -4542,17 +4550,23 @@ bool ChainstateManager::ProcessNewBlock(const std::shared_ptr<const CBlock>& blo
     BlockValidationState state; // Only used to report errors, not invalidity - ignore it
     if (!ActiveChainstate().ActivateBestChain(state, block)) {
         LogError("%s: ActivateBestChain failed (%s)\n", __func__, state.ToString());
-        return false;
+        if (state.IsValid()) NONFATAL_UNREACHABLE();
+        return state;
     }
 
     Chainstate* bg_chain{WITH_LOCK(cs_main, return BackgroundSyncInProgress() ? m_ibd_chainstate.get() : nullptr)};
-    BlockValidationState bg_state;
-    if (bg_chain && !bg_chain->ActivateBestChain(bg_state, block)) {
-        LogError("%s: [background] ActivateBestChain failed (%s)\n", __func__, bg_state.ToString());
-        return false;
-     }
 
-    return true;
+    // Attempts to connect the block to the tip of the active chain.
+    BlockValidationState activation_state;
```

Comment:
I'm confused by this change. Why are you introducing this variable? The comment is also wrong, since this processes the block on the background chain. I would still prefer if the behaviour of ProcessNewBlock would not be changed as part of this patch set.

Reply:
The comment you linked seems misleading. `bg_state` is the correct name in my view. We are not operating on the active chain here.

I think the belt-and-suspenders check in `ActivateBestChain` is broken by this change. When we return false in the case of the chain being disabled, we now run into `NONFATAL_UNREACHABLE`. It might be true that there are no similar cases in the call graph of `ActivateBestChain`, but can we guarantee that? For this reason, I think the commit should be dropped.